kadi/events/query.py

```diff
@@ -6,6 +6,9 @@
 
 from Chandra.Time import DateTime
 
+import django
+django.setup()
+
```

Comment:
This is the magic that makes django work.

This being here causes an error if one imports kadi.events.models befare kadi.events.query:
```python
In [1]: from kadi.events import models
---------------------------------------------------------------------------
AppRegistryNotReady                       Traceback (most recent call last)
<ipython-input-1-0caab80e9759> in <module>
----> 1 from kadi.events import models

~/git/kadi/kadi/events/models.py in <module>
    239 
    240 
--> 241 class Update(models.Model):
    242     """
    243     Last telemetry which was searched for an update.

/export/servers/kadi-test/ska3/lib/python3.8/site-packages/django/db/models/base.py in __new__(cls, name, bases, attrs, **kwargs)
    106 
    107         # Look for an application configuration to attach the model to.
--> 108         app_config = apps.get_containing_app_config(module)
    109 
    110         if getattr(meta, 'app_label', None) is None:

/export/servers/kadi-test/ska3/lib/python3.8/site-packages/django/apps/registry.py in get_containing_app_config(self, object_name)
    251         Return None if the object isn't in any registered app config.
    252         """
--> 253         self.check_apps_ready()
    254         candidates = []
    255         for app_config in self.app_configs.values():

/export/servers/kadi-test/ska3/lib/python3.8/site-packages/django/apps/registry.py in check_apps_ready(self)
    134             # exception.
    135             settings.INSTALLED_APPS
--> 136             raise AppRegistryNotReady("Apps aren't loaded yet.")
    137 
    138     def check_models_ready(self):

AppRegistryNotReady: Apps aren't loaded yet.

In [2]: from kadi.events import query

In [3]: from kadi.events import models
```

Reply:
Good catch. It's like whack-a-mole. I'm just getting very annoyed with django, but the whole kadi events are pretty deeply integrated into that framework.

At some level users don't need to import `kadi.events.model` (or could import query first) but I agree it has a bad taste.

Reply:
Well I was hitting the server for a while with no errors and feeling good, until:

{"error": "Traceback (most recent call last):\n  File \"/export/servers/config/test/wsgi-scripts/api.py\", line 66, in application\n    app_func, app_kwargs, server_opt = parse_url(environ)\n  File \"/export/servers/config/test/wsgi-scripts/api.py\", line 142, in parse_url\n    app_func = getattr(app_func, func_part)\n  File \"/export/servers/kadi-test/ska3/lib/python3.8/site-packages/kadi/events/__init__.py\", line 84, in __getattr__\n    query = importlib.import_module('kadi.events.query')\n  File \"/export/servers/kadi-test/ska3/lib/python3.8/importlib/__init__.py\", line 127, in import_module\n    return _bootstrap._gcd_import(name[level:], package, level)\n  File \"<frozen importlib._bootstrap>\", line 1014, in _gcd_import\n  File \"<frozen importlib._bootstrap>\", line 991, in _find_and_load\n  File \"<frozen importlib._bootstrap>\", line 975, in _find_and_load_unlocked\n  File \"<frozen importlib._bootstrap>\", line 671, in _load_unlocked\n  File \"<frozen importlib._bootstrap_external>\", line 843, in exec_module\n  File \"<frozen importlib._bootstrap>\", line 219, in _call_with_frames_removed\n  File \"/export/servers/kadi-test/ska3/lib/python3.8/site-packages/kadi/events/query.py\", line 10, in <module>\n    django.setup()\n  File \"/export/servers/kadi-test/ska3/lib/python3.8/site-packages/django/__init__.py\", line 24, in setup\n    apps.populate(settings.INSTALLED_APPS)\n  File \"/export/servers/kadi-test/ska3/lib/python3.8/site-packages/django/apps/registry.py\", line 114, in populate\n    app_config.import_models()\n  File \"/export/servers/kadi-test/ska3/lib/python3.8/site-packages/django/apps/config.py\", line 211, in import_models\n    self.models_module = import_module(models_module_name)\n  File \"/export/servers/kadi-test/ska3/lib/python3.8/importlib/__init__.py\", line 127, in import_module\n    return _bootstrap._gcd_import(name[level:], package, level)\n  File \"<frozen importlib._bootstrap>\", line 1014, in _gcd_import\n  File \"<frozen importlib._bootstrap>\", line 991, in _find_and_load\n  File \"<frozen importlib._bootstrap>\", line 975, in _find_and_load_unlocked\n  File \"<frozen importlib._bootstrap>\", line 671, in _load_unlocked\n  File \"<frozen importlib._bootstrap_external>\", line 843, in exec_module\n  File \"<frozen importlib._bootstrap>\", line 219, in _call_with_frames_removed\n  File \"/export/servers/kadi-test/ska3/lib/python3.8/site-packages/kadi/events/models.py\", line 9, in <module>\n    django.setup()\n  File \"/export/servers/kadi-test/ska3/lib/python3.8/site-packages/django/__init__.py\", line 24, in setup\n    apps.populate(settings.INSTALLED_APPS)\n  File \"/export/servers/kadi-test/ska3/lib/python3.8/site-packages/django/apps/registry.py\", line 83, in populate\n    raise RuntimeError(\"populate() isn't reentrant\")\nRuntimeError: populate() isn't reentrant\n"}

Reply:
I guess you either already know this or are googling it right now...

https://blog.whtsky.me/tech/2020/populate_isnt_reentrant

Reply:
In my local copy I just added the call to `django.setup()` inside `kadi.models`. I don't know why it worked fine for a while.

Reply:
Yes, I've been there, done that.. :smile:  Probably more useful is the first answer in this SO: https://stackoverflow.com/questions/27093746/django-stops-working-with-runtimeerror-populate-isnt-reentrant.

The problem is that the issue only shows up on the Apache web server. So maybe we could monkeypatch the web server django to get a better traceback?

About adding `django.setup()` inside `kadi.models`, this will break `manage.py` because that calls `setup()` and then imports `kadi.models` and you reliably get the reentrant exception. `manage.py` is still required in order to build or migrate the database models.

In any case, my best effort of having only one place in the kadi code that does `django.setup()` has failed to stop the reentrant problem. One guess is that this is related to Apache running Python sub-interpreters so that something really funky is happening.

My only answer now is to just punt on the kadi events web-kadi API service. AFAIK there are zero actual users. We test it a lot but I don't think it is used for actual work.

Reply:
as to where to put the call to `django.setup()`, you can have it inside a little module that only does that and is imported from all others, essentially letting python do the lazy load, depending on which module is imported first. That would not prevent funky business with WSGI sub-interpreters but would fix our usual use cases (importing `query`, `models` or running `manage.py`.

This would still be the point where we would have crazy logic to try to prevent double loading and such.

Reply:
@javierggt - I pushed a new commit to catch the reentrant exception. Can you try installing that to web-kadi-test?

Reply:
I removed the change I made to add the call to setup inside `models`. That might actually have been the cause of the error you found.

There are two tests I can make before monkeypatching django in kadi-test:
- make sure the top-level app imports `kadi.event.query`
- set the process group to `%{GLOBAL}`